java-security/src/main/java/com/sap/cloud/security/token/HybridTokenFactory.java

```diff
@@ -14,44 +13,33 @@
 
 /**
  * Creates a {@link Token} instance. Supports Jwt tokens from IAS and XSUAA
- * identity service. This will be moved to java-api component soon and will load
- * and instantiate the respective Token dynamically.
+ * identity service. TokenFactory loads and instantiates the respective Token
+ * dynamically.
  */
-class TokenFactory {
+public class HybridTokenFactory implements TokenFactory {
 
-	private TokenFactory() {
+	public HybridTokenFactory() {
 		// use the factory method instead
 	}
 
 	/**
-	 * Determines whether the JWT token is issued by XSUAA identity service, and
-	 * creates a Token for it.
+	 * Determines whether the JWT token is issued by XSUAA or IAS identity service,
+	 * and creates a Token for it.
 	 *
 	 * @param jwtToken
 	 *            the encoded JWT token (access_token or id_token), e.g. from the
-	 *            Authorization Header.
+	 *            Authorization Header. the scope converter, e.g.
+	 *            {@link XsuaaScopeConverter}
+	 * @param xsuaaAppId
+	 *            Xsuaa application Id from the CF environment
 	 * @return the new token instance
 	 */
-	public static Token create(String jwtToken) {
-		return create(jwtToken, null);
-	}
-
-	/**
-	 * Determines whether the JWT token is issued by XSUAA identity service, and
-	 * creates a Token for it.
-	 *
-	 * @param jwtToken
-	 *            the encoded JWT token (access_token or id_token), e.g. from the
-	 *            Authorization Header.
-	 * @param localScopeConverter
-	 *            the scope converter, e.g. {@link XsuaaScopeConverter}
-	 * @return the new token instance
-	 */
-	public static Token create(String jwtToken, ScopeConverter localScopeConverter) {
+	public Token create(String jwtToken, String xsuaaAppId) {
 		Objects.requireNonNull(jwtToken, "Requires encoded jwtToken to create a Token instance.");
 		DecodedJwt decodedJwt = Base64JwtDecoder.getInstance().decode(removeBearer(jwtToken));
 
 		if (isXsuaaToken(decodedJwt)) {
+			ScopeConverter localScopeConverter = new XsuaaScopeConverter(xsuaaAppId);
```

Comment:
Hi @liga-oz 

this change looks now very pretty... beside of this... xsuaa appId is not a runtime information, means it does not change during runtime. That means HybridTokenFactory should create only once the ScopeConverter and for that it need to fetch the appId from the environment (which is cached)...

see also here:
https://github.com/SAP/cloud-security-xsuaa-integration/blob/master/java-security/src/main/java/com/sap/cloud/security/servlet/XsuaaTokenAuthenticator.java#L47